### ISSUE TITLE
Proposal: Extend NetworkPolicy to Layer 7

### ISSUE THREAD

**Issue by tgraf** [kubernetes/kubernetes#58624]
/kind feature
/sig network

# Extend NetworkPolicy to Layer 7
## Objective
This proposal extend NetworkPolicy to include Layer 7 attributes at both ingress and egress to extend filtering on HTTP method, HTTP path, Kafka topic, etc.

## Motivation
The current NetworkPolicy allows to apply security policies based on L3
and L4 information (pod to pod connectivity, pod to IP, port filtering). Many applications running in Kubernetes pods use modern protocols such as gRPC or REST to expose APIs. Securing such pods on pod and port level only provides limited security as the entire API is exposed. This proposal provides users the capability to lock down pods on API call level.

## Goals
* Provide a L7/API aware security specification compatible with NetworkPolicy
* The goal is to provide enforcement for:
   * Traffic that is coming from a pod (egress)
   * Traffic that is going to a pod (ingress)
* Find a simple language to express the most common use cases in locking down on L7/API level

## Non-Goals
* This proposal does not intent to capture L7 routing intent and is not a replacement for Ingress
* It is not the goal to provide capability to filter on every possible L7 protocol field.

### API changes
We introduce a new field `Rules` on the existing `NetworkPolicyPort` structure which will contain rules for traffic passing through that port.

```go
// NetworkPolicyPort describes a port to allow traffic on
type NetworkPolicyPort struct {
	// The protocol (TCP or UDP) which traffic must match. If not specified, this
	// field defaults to TCP.
	// +optional
	Protocol *v1.Protocol `json:"protocol,omitempty" protobuf:"bytes,1,opt,name=protocol,casttype=k8s.io/api/core/v1.Protocol"`

	// The port on the given protocol. This can either be a numerical or named port on
	// a pod. If this field is not provided, this matches all port names and numbers.
	// +optional
	Port *intstr.IntOrString `json:"port,omitempty" protobuf:"bytes,2,opt,name=port"`

+	// The rule on the given port/protocol tuple. If this field is
+	// provided, the Port and Protocol field must be set as well and traffic will only be
+      // allowed when the traffic matches the provided rules.
+      // 
+      // +optional
+      Rules L7PortRule `json:"rules,omitempty" protobuf:"bytes,3,opt,name=application"`
}

// L7PortRule is a union of port level L7 rules. Mixing of different port
// level rule types is disallowed, so exactly one of the following must be set.
// If none are specified, then no additional port level rules are applied.
type PortRule struct {
	// HTTP specific rules.
	//
	// +optional
	HTTP []PortRuleHTTP `json:"http,omitempty" protobuf:"bytes,1,opt,name=http"`


         // Kafka specific rule.
         //
         // +optional
	Kafka []PortRuleKafka `json:”kafka,omitempty” protobuf:”bytes,2,opt,name=kafka”`
}
// PortRuleHTTP is a list of HTTP protocol constraints. All fields are
// optional, if all fields are empty or missing, the rule does not have any
// effect.
//
// All fields of this type are extended POSIX regex as defined by IEEE Std
// 1003.1, (i.e this follows the egrep/unix syntax, not the perl syntax)
// matched against the path of an incoming request. Currently it can contain
// characters disallowed from the conventional "path" part of a URL as defined
// by RFC 3986.
type PortRuleHTTP struct {
	// Path is an extended POSIX regex matched against the path of a
	// request.
	// If omitted or empty, all paths are all allowed.
	//
	// +optional
	Path string `json:"path,omitempty" protobuf:"bytes,1,opt,name=path"`
	// Method is an extended POSIX regex matched against the method of a
	// request, e.g. "GET", "POST", "PUT", "PATCH", "DELETE", ...
	//
	// If omitted or empty, all methods are allowed.
	//
	// +optional
	Method string `json:"method,omitempty" protobuf:"bytes,2,opt,name=method"`
	// Host is an extended POSIX regex matched against the host header of a
	// request, e.g. "foo.com"
	//
	// If omitted or empty, the value of the host header is ignored.
	//
	// +optional
	Host string `json:"host,omitempty" protobuf:"bytes,3,opt,name=host"`
	// Headers is a list of HTTP headers which must be present in the
	// request. If omitted or empty, requests are allowed regardless of
	// headers present.
	//
	// +optional
	Headers []string `json:"headers,omitempty" protobuf:"bytes,4,opt,name=headers"`
}

// PortRuleKafka is a list of Kafka protocol constraints. All fields are
// optional, if all fields are empty or missing, the rule will match all
// Kafka messages.
type PortRuleKafka struct {
        // APIVersion is the version matched against the api version of the
        // Kafka message. If set, it has to be a string representing a positive
        // integer.
        //
        // If omitted or empty, all versions are allowed.
        //
        // +optional
        APIVersion string `json:"apiVersion,omitempty"`
        // APIKey is a case-insensitive string matched against the key of a
        // request, e.g. "produce", "fetch", "createtopic", "deletetopic", et al
        // Reference: https://kafka.apache.org/protocol#protocol_api_keys
        //
        // If omitted or empty, all keys are allowed.
        //
        // +optional
        APIKey string `json:"apiKey,omitempty"`
        // ClientID is the client identifier as provided in the request.
        //
        // From Kafka protocol documentation:
        // This is a user supplied identifier for the client application. The
        // user can use any identifier they like and it will be used when
        // logging errors, monitoring aggregates, etc. For example, one might
        // want to monitor not just the requests per second overall, but the
        // number coming from each client application (each of which could
        // reside on multiple servers). This id acts as a logical grouping
        // across all requests from a particular client.
        //
        // If omitted or empty, all client identifiers are allowed.
        //
        // +optional
        ClientID string `json:"clientID,omitempty"`
        // Topic is the topic name contained in the message. If a Kafka request
        // contains multiple topics, then all topics must be allowed or the
        // message will be rejected.
        //
        // This constraint is ignored if the matched request message type
        // doesn't contain any topic. Maximum size of Topic can be 249
        // characters as per recent Kafka spec and allowed characters are
        // a-z, A-Z, 0-9, -, . and _
        // Older Kafka versions had longer topic lengths of 255, but in Kafka 0.10
        // version the length was changed from 255 to 249. For compatibility
        // reasons we are using 255
        //
        // If omitted or empty, all topics are allowed.
        //
        // +optional
        Topic string `json:"topic,omitempty"`
}
```
## Examples 
1) A policy that only allows traffic from pods with labels `env=frontend` to
pods with labels `env=backend` that is a HTTP GET request to path `/index.html`
on port 80/TCP:
```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: with-l7-policy
spec:
  podSelector:
    matchLabels:
      env: backend
  ingress:
  - from:
    - podSelector:
        matchLabels:
          env: frontend
    ports:
    - protocol: TCP
      port: 80
      rules:
        http:
          path: "\/index\.html"
          method: "GET"
```
2) A policy that only allows traffic from pods with labels `env=frontend` to
pods with labels `env=backend` that is a HTTP POST request with the prefix
`/path1/.*` on port 80/TCP or a HTTP GET request with the prefix
`/path2/.*` on port 8080.
```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: with-l7-policy-to-2-ports
spec:
  podSelector:
    matchLabels:
      env: backend
  ingress:
  - from:
    - podSelector:
        matchLabels:
          env: frontend
    ports:
    - protocol: TCP
      port: 80
      application:
        http:
          path: "\/path1\/.*"
          method: "POST"
    - protocol: TCP
      port: 8080
      rules:
        http:
          path: "\/path2\/.*"
          method: "GET"
```
4) A policy that only allows “produce” but no “consume” traffic from “kafka-producer” to “kafka” on port 9092 TCP.
```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: l7-kafka
spec:
  podSelector:
    matchLabels:
      env: kafka
  ingress:
  - from:
    - podSelector:
        matchLabels:
          env: kafka-producer
    ports:
    - protocol: TCP
      port: 9092
        rules:
          kafka:
          - apiKey: "apiversions"
          - apiKey: "metadata"
          - apiKey: "produce"
            topic: "deathstar-plans"
          - apiKey: "produce"
            topic: "empire-announce"
```

**Comment by danwinship** [comment]
This seems WAY out of scope for NetworkPolicy to me. Kubernetes defines an L3/L4 network, so the network plugin needs to be aware of L3/L4 things, so being able to filter on L3/L4 attributes is "easy" and logical. A kubernetes network plugin does not inherently need to have any awareness of L7 things, and AFAIK the majority of currently-existing ones don't.

**Comment by sbezverk** [comment]
@tgraf I find it a little odd to add new type for a specific application, would it not make more sense to provide a set of L7 related primitives to be able to construct any L7 policies?
When kafka changes something in their API, there will be a pressure from Kafka's user community to apply similar changes to k8s as well. 
I guess similar approach was taken in storage where there is a persistent volume and the storage backend specifics in a persistent volume source, but the difference is the number of storage backends is 10-20, number of applications is skyrocketing.

**Comment by tgraf** [comment]
> would it not make more sense to provide a set of L7 related primitives to be able to construct any L7 policies?

I don't really like the concept of multiple resources referring to each other unless there is a clear n:1 relationship going on.

> When kafka changes something in their API, there will be a pressure from Kafka's user community to apply similar changes to k8s as well.

Happy to discuss specifics around Kafka. The goal is *not* to support everything the Kafka protocol supports. It is limited to the base header type which has been stable since the very beginning plus the topic name which is a concept in Kafka that has been unchanged from the beginning. That said, if Kafka is extended and it makes sense to extend the policy spec along with it and enough users backing it, I don't see an issue doing that.

> I guess similar approach was taken in storage where there is a persistent volume and the storage backend specifics in a persistent volume source, but the difference is the number of storage backends is 10-20, number of applications is skyrocketing.

The proposal explicitly defines supporting all application protocols as a non goal. Can you list more than 20 more protocols that you want to support? There are many other protocols besides IP, TCP and UDP, yet NetworkPolicy decided to only support those and ignore any L2 protocol filtering, Non-IP, ICMP, SCTP, ...

**Comment by tgraf** [comment]
> This seems WAY out of scope for NetworkPolicy to me. Kubernetes defines an L3/L4 network, so the network plugin needs to be aware of L3/L4 things, so being able to filter on L3/L4 attributes is "easy" and logical. A kubernetes network plugin does not inherently need to have any awareness of L7 things, and AFAIK the majority of currently-existing ones don't.

Given the interest and traction of Istio and service mesh, L7 will likely quickly be the norm instead of the exception. It can either be part of NetworkPolicy or not. This proposal kicks off the discussion.

We already have three different CRDs from major plugins (2 CNI + Istio) involving L7. It will definitely happen, it's just a matter of how.

I would also decouple the question of whether all of NetworkPolicy has to be implemented by each plugin. I believe we already discussed that question very early on with the answer being no.

If not part of NetworkPolicy, the interesting questions to me are:
- How do we enable source dependent L7 policies, e.g. foo consuming bar can do `GET /foo` and blah consuming bar can do `GET /bar`.
- How do we map ports to application protocol rules without completely duplicating the L4 policy part in two different resource files?

**Comment by rshriram** [comment]
drive by comment:
How about a generic l7 directive like `type.resource.operation` to denote all forms of l7 stuff? messagebus.topic.publish/subscribe, db.<table>.read/write, http.<service/dns/regex>.get/put/post, etc.. ? Kubernetes does not have to know or interpret the three fields.. Its an opaque string with 3 elements, leaving it to the implementations to interpret. They can ignore things they don't support.

**Comment by louiscryan** [comment]
I generally agree with @tgraf about the direction the market is heading. I do wonder how & where Kubernetes wants to set its functional boundaries around features like this. 

The Istio approach has been to codify the extraction of traffic properties and then provide an integration point for engines to implement their custom magic. This approach would make more sense to me if K8S had a mechanism or standing commitment to cover more and more protocols as first-class in its APIs. I'm a little skeptical of that as a goal and I suspect others are too

**Comment by elevran** [comment]
> Given the interest and traction of Istio and service mesh, L7 will likely quickly be the norm instead of the exception. It can either be part of NetworkPolicy or not. This proposal kicks off the discussion.

Agree that service meshes will drive interest here. 
I don't think it belongs in a k8s NetworkPolicy. An alternative may be to (optionally) derive NetworkPolicy objects from a higher level, application-centric, policy. 

Also, the use of having protocol specific entities in the spec may not scale. Agree with @louiscryan and @rshriram about trying to consolidate and generalize. 
Having `operation`, `resource` and `path` might address a large part of customers' requirements. 
Drawing an (inaccurate?) analogy from RBAC, we already have this defined:

```yaml
rules:
- apiGroups: [""]
  resources: ["pods", "pods/log"]
  verbs: ["get", "list"]
```

> How do we map ports to application protocol rules without completely duplicating the L4 policy part in two different resource files?

The current port mapping associates a name with a port number. Unless we mandate and enforce some specific naming conventions, there might not be a relation between the port name and the application protocol that runs over it. I have a feeling this will require a separate mapping anyway, one that can be part of the application level policy definition.

**Comment by tgraf** [comment]
> I generally agree with @tgraf about the direction the market is heading. I do wonder how & where Kubernetes wants to set its functional boundaries around features like this.

Yes, this is exactly the purpose of this issue. In my opinion, Kubernetes either includes L7 or policy becomes obsolete in Kubernetes. I doubt that users will want to duplicate allowed service to service connectivity and port numbering information across two different systems.

> The Istio approach has been to codify the extraction of traffic properties and then provide an integration point for engines to implement their custom magic. This approach would make more sense to me if K8S had a mechanism or standing commitment to cover more and more protocols as first-class in its APIs. I'm a little skeptical of that as a goal and I suspect others are too

The assumption is definitely to cover the most common protocols. I think this is not much different than what happened to deployment logic. A couple of years ago, all deployment was codified and everyone had a bunch of scripts around container runtimes. Believing the CNCF user reports, bash is still a heavy orchestrator out there. Kubernetes has standardised this and found the right level of capturing the user intent. I believe that covering the 5-10 most common protocols will cover the magic 80%. I also don't think that one model excludes the other.

**Comment by lxpollitt** [comment]
> Given the interest and traction of Istio and service mesh, L7 will likely quickly be the norm instead of the exception. It can either be part of NetworkPolicy or not. This proposal kicks off the discussion.

My view is that unified security policy rather than separate Kubernetes and Istio policies is valuable for many users, and that Kubernetes Network Policy idioms are a solid base for this policy model.  (Perhaps no surprise that both Cilium and Calico are experimenting in this area!)

There's a big part of me that would like to drive wider acceptance of basing L7 policy around Kubernetes Network Policy idioms. However, I'm hesitant about trying to codify this into the Kubernetes Network Policy definition.  I think that the reality is that for most network implementations L7 is way out of scope (as @danwinship suggested).  

> In my opinion, Kubernetes either includes L7 or policy becomes obsolete in Kubernetes.

I'm not convinced they will be obsolete.  Istio is getting some great traction but it won't be for everyone.  We've had plenty of L7 connectivity evolutions before (e.g. ESB, etc), and networking level (L3-4) security remained relevant for most users.  But equally, repeating myself, I do think there's a set of users for unified L3-7 policy is going to be a valuable feature.

**Comment by lxpollitt** [comment]
> I believe that covering the 5-10 most common protocols will cover the magic 80%

We kept the protocols that Network Policy recognizes today pretty small, but I feel less confident whether we could achieve a stable set of 5-10 L7 protocols.  There's a lot of L7 protocols!  So feels like either we will be forever growing this list due to pressure from a particular vendor/project/large user, or we will be falling back to annotations or CRDs as an extension mechanism to cover protocols not in the list. I was talking with a large enterprise today and they had a list of 15 protocols they regarded as essential if they were going to move to an L7 service mesh!

**Comment by lxpollitt** [comment]
Ideally there would be a better way of extending Kubernetes API objects so we could define experimental extensions or vendor/project specific extensions.  Annotations are super limited mechanism for doing this for API objects that have significant structure to them such as Network Policy.  You really want to be able to add extension fields anywhere within the existing structure.  I know this topic has come up a few times, but I don’t know the latest thinking or if there is any API work underway that is considering addressing this?  

We could maybe add a per rule annotation/extension field to the Kubernetes Network Policy definition?  But that's might prove pretty contentious to get agreement on.

Perhaps in the short term it would be best to follow a CRD approach.  e.g. Vendor/project specific CRDs that mirror the Kubernetes Network Policy structure, but with the addition of any extensions the project supports.  It keeps the contract with the user clean and it allows the project to move at its own velocity without dependency on broader community agreement.  In time we'll have more user validation of the extensions and can perhaps build a stronger case for bringing into core Kubernetes at that point.  (This is basically the approach we've used in Calico to date with other network policy features, rather than trying to push for every Calico feature to be included in the core Kubernetes Network Policy API.)

**Comment by rikatz** [comment]
/reopen
/remove-lifecycle rotten
/assign @rikatz @jayunit100 @abhiraut 
Let's tackle this in 2021 and see how viable is this for netpol evolutions.

**Comment by aojea** [comment]
> /reopen
> /remove-lifecycle rotten
> /assign @rikatz @jayunit100 @abhiraut
> Let's tackle this in 2021 and see how viable is this for netpol evolutions.

This is more in line with the service-apis group work that already have the primitives to deal with applications https://kubernetes-sigs.github.io/service-apis/  

@bowei are you already working on something of this or do you have it in your roadmap?

**Comment by rikatz** [comment]
/lifecycle frozen
We're coming back to this on Q3/2021, as almost all the group is out of bandwidth for taking a look into this, unless someone wants to join us to discuss this one :)

**Comment by thockin** [comment]
Accepting this - not committing to DOING it, but it doesn't need triage

**Comment by tgraf** [comment]
I'm happy to get this to a concrete proposal with a KEP based on what we have learned with CiliumNetworkPolicy.

**Comment by rikatz** [comment]
@tgraf one thing we've been discussing in Network Policy v2 is if we want L7 to be part of Network Policy, or if Netpol will be only related to L3/L4, and maybe L7 be other object like AppPolicy, TrafficPolicy, etc.

Doing L3/L4 related is easier for CNI providers to adopt without too much caveat, but still allowing the extensibility of the Network Policy object to selectors that ends on a combination of IP/Port for each object.

OTOH, working with L7 policies imply on CNIs having to run some sidecar (as an example) and directing the traffic to it to allow or deny based on Methods (as Cilium already does) and this may brings some more complexity to implementations :)

We've started actually to map the premises of the evolution of NetPol, feel free to reach us on the meetings (or Slack, etc!) and we can further discuss about this!

**Comment by jayunit100** [comment]
i think abhishek and others are looking at this
/unassign

**Comment by thockin** [comment]
This old issue was revived by the bot.

We discussed today at sig-net whether / how k8s proper could take on this problem.

The consensus points I took away were:

1) We should probably indicate where such policy rules would go in the Gateway API.
2) We probably don't need to own all the expressions of such policy - L7 could mean HTTP (which can get arbitrarily complicated) but it could also mean MySQL or Kafka or Redis.  I don't think k8s should own all of those.  Gateway already owns *some* policy-ish definitions for HTTP, and it seems reasonable that *very common* ones find their way into that set.

This seems like a place where the extensible core + ecosystem implementations will be a better solution.

Given that, we're closing this issue.
